[wip] [ember-data] fix references API return types

ro0gr opened the pull request:
This enables proper return types for model references API, so assuming the following example:

```ts
export default class User extends Model {
  @belongsTo("user")
  friend!: AsyncBelongsTo<User>;
}

// ... 

const user = store.createRecord('user');

const friend = user.belongsTo('friend').value() 
```

The `friend` now has `User | null` type. And prior to this change it had the `Model | null` type, which did often require to specify an exact return type explicitly, like:

```ts
const friend = user.belongsTo('friend').value() as User|null;
```

As a bonus, it does also fix `this` when used with the references API(https://github.com/typed-ember/ember-cli-typescript/issues/1416), due to a change of how we handle `this`:

```ts
// before
belongsTo(name: RelationshipsFor<this>): ...;

// after
belongsTo<T extends Model, K extends keyof BelongsToModels<T>>(this: T, name: K): ...;
```

This implementation is restored from the https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42839. The only notiecable difference is that now we do only properly resolve references for the properties whose types are declared as `AsyncHasMany`, `SyncHasMany`, `AsyncBelongsTo`, and `Model`. Properties declared with the "old" types, like `DS.PromiseObject`, `DS.PromiseManyArray`, which was respected in the previous PR, are ignored now, and it's supposed to work w/o changes. I believe this way we can stay a kind of backward compatible, and encourage folks to migrate to the new convenience types at the same time! However, I'm open to add the "old" types handling, please let me know if I should to.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

sandersn commented:
This is a complex change, so I don't want to merge it without a review from one of the owners. @dwickern @mike-north @chriskrycho @jamescdavis @dfreeman can you take a look? I can take another look to give technical advice on the structure of the types if you need me to, but I don't know much about ember.

mike-north commented on types/ember-data/index.d.ts:
```diff
@@ -742,7 +769,7 @@ export namespace DS {
      * relationship.
      */
-    class BelongsToReference {
+    class BelongsToReference<T extends Model> {
```
To avoid making this a breaking change for users who already consume the `BelongsToReference` type, I think we need a default type param `= Model`

mike-north commented on types/ember-data/test/belongs-to.ts:
```diff
@@ -29,5 +30,6 @@ folder.set('parent', folder);
-// $ExpectType Model | null
-folder.belongsTo('parent').value();
+folder.belongsTo('parent').value(); // $ExpectType Model | null
+folder.belongsTo('parentSync').value(); // $ExpectType Model | null
```
Can we introduce a test case for a model with no relationships, just to make sure the `never` case collapses down properly?

ro0gr replied:
Honestly, I'm not sure about the

> `never` case collapses down properly

if you mean, making sure that it should not be possible to pass any argument into the `belongsTo`, if there are no any relationships defined on a model.. It seems impossible to do from my experience due to the keys filtering issue described above.

As far as I can see, the best we can do now is to turn `belongsTo(`, and `hasMany(` to return `never` in case of invalid relationship key.

ro0gr replied:
Honestly I'm not exactly sure about the

> `never` case collapses down properly

do you mean, to make it impossible to pass any key to the `belongsTo(`/`hasMany(` if there are no any relationships defined on a model? If such, from [my experience](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51852#discussion_r603631700) it seems impossible to do currently.

I think the best we can achieve here for now is to turn `belongsTo(`/`hasMany(` to return `never` for invalid keys, but not fail the build unfortunately.

mike-north commented on types/ember-data/index.d.ts:
```diff
@@ -25,7 +25,30 @@ import AdapterRegistry from 'ember-data/types/registries/adapter';
+      : DS.Model;
+
+type BelongsToModels<Model extends DS.Model> = {
+    [k in keyof Model]: UnboxBelongsTo<Model[k]>
```
Purely a stylistic preference, but can we make `k` a capital letter?

 This helps to disambiguate between constrained and unconstrained index signatures (e.g., `{ [k: string]: number }` where `k` is not a type, and `{ [K in keyof Model]: string }` where `K` is a type)

mike-north commented on types/ember-data/index.d.ts:
```diff
@@ -25,7 +25,30 @@ import AdapterRegistry from 'ember-data/types/registries/adapter';
+      : T extends DS.SyncHasMany<infer C> ? C
+      : T extends DS.AsyncBelongsTo<any> ? never
+      : T extends DS.Model ? never
+      : DS.Model;
```
I haven't had a chance to run this locally, but the `DS.Model` fallback in the final condition here makes me wonder whether `BelongsToModels<Model>` will turn something like
```ts
interface Model {
   nonPersistedField1: string;
   nonPersistedField2: string;
   nonPersistedField3: string;
   relatedThings: DS.AsyncHasMany<Foo>
}
```
into 
```ts
interface Model {
   nonPersistedField1: DS.Model;
   nonPersistedField2: DS.Model;
   nonPersistedField3: DS.Model;
   relatedThings: Foo
}
```
Can we introduce some more test cases that make it clear that this utility type works the way you want it to?

I'm particularly interested in cases like

* `DS.attr()` (no property type)
* a regular (not `DS.attr` )string class field that's present on an e-d model class)
* Models with no relationships to other models at all
* Models with computed properties

some of the patterns above are inadvisable, but they are supported and exist in the real world. We should make sure that they don't _break_ with this change.

ro0gr replied:
That's how it works currently, and this PR doesn't change this behavior, as confirmed by a test case( [before](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51852/commits/9403bc13b7184ffaf3b05c34b4dae74365dfdaa3#diff-6fd426c31e33023aa19e22cae72a4f94d106a4d3575ebffe9f77260406ee6f15R116), [after](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51852/files#diff-6fd426c31e33023aa19e22cae72a4f94d106a4d3575ebffe9f77260406ee6f15R115) ) written by your suggestion.

I'm open to also fix this issue, and start respecting `DS.PromiseObject`, `DS.PromiseManyArray`, and the rest from the  [my previous attempt](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42839/files#diff-f84a1c4948e4f7317ca9d56ad8a88c9ae199d5fb9cb327ae8d1add268f2c725dR30-R39). Please let me know if I should do it?

cc @jamescdavis

mike-north reviewed (changes requested):
Thanks for working on this change! I think what we need to carry this forward is
- [ ] Additional tests (specifics of tests I'd like to see are in review comments)
- [ ] We need someone more familiar with the ember-data references API to review this
- [ ] Any new type params we introduce should either be _always inferred_ or should have default types `<T extends Model = Model>` to avoid introducing unnecessary breaking changes. (we should include tests that _prove_ we remain compatible)

Hopefully you have some actionable feedback to respond to @ro0gr -- I only have intermittent availability, but will do my best to keep an eye on this. Feel free to `@mention` me when you'd like a re-review

ro0gr commented:
@mike-north thank you for your review! There are many great points! Will try to address it this week.

ro0gr reviewed (commented):
@mike-north I believe I addressed what is possible to address, and left some explanations for the current problems. Please let me know what you do think if you have a chance to take another look.

ro0gr commented:
@jamescdavis I'd really appreciate if you take a look here!

ro0gr commented:
Oh.. I've just applied this patch on my test suite, and seems there are some breakages. going to investigate and come back with the fixes.

ro0gr commented:
there are definitely some breaking changes with this PR. I'll come back once I wrap it up, and share the details.

ro0gr commented:
So I've detected some issues in the codebase I'm working on, and addressed them in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51852/commits/6fd2b010424a8b48d8f1168065a0e2fe6b21ee88, and https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51852/commits/de3b7c982da8ee2e8eda800375f62bef22ca328d.

Here are my notes regarding changes that was needed to do in my codebase, in order to apply this PR:

### get rid of exlicit `this` in accesor/method signatures

In my case, most ocurrences of `this: Anything` just worked. But sometimes when it came to the bi-directional relationships in conjunction with an explicit `this`, it led to a circular reference error of the typescript. So I just found all the ocurrences of the `(this: `, and removed each. I think, taking into account https://github.com/typed-ember/ember-cli-typescript/issues/1416 this kind of cleanup is a good thing in general.

### Array | ArrayProxy | EmberArray

Prior to this PR, any property of a model(excluding the properties of the `DS.Model`) was allowed to be a relationship, so you could:
```ts
  myModel.hasMany('title'); // $ExpectType HasManyReference<DS.Model>
```
which allowed us to mark sync has manies, with invalid types, and I found some ocurrences like `Array`, `ArrayProxy`, or `EmberArray` in my code.

Now, for `hasMany` we do only recognize `AsyncHasMany`, `SyncHasMany`, and return type of the `DS.hasMany(`. So after upgradiung the types all the usages of `hasMany` agains `Array`, `ArrayProxy`, or `EmberArray` lead to error.

IMO, it should not take too much to update in the userland to a proper hasMany types, but from the other hand it could be called a breaking change. 

Please let me know if I should update typings to consider these 3 array kinds as `SyncHasMany`.

ro0gr commented:
solved the conflicts caused by https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52241, but I expect CI to fail now due to the https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52240#discussion_r610097489. Hopefuly we'd find a good path forward for this.

chriskrycho commented:
@wagenet @ro0gr @jamescdavis can you three put your heads together and ID the right shape of a solution here? 🙏🏼

ro0gr commented:
@chriskrycho thanks for the heads up! 

I think, I need to re-evaluate my understanding of the current state of things. Going to come back to it this weekend.

wagenet commented:
@ro0gr sorry I have had a chance to give this a look yet. Feel free to reach out to me on Discord as well.

ro0gr commented:
CI seems to pass now. However, I've found another set of new issues in my codebase. Marking this as a WIP for now.

ro0gr commented:
As usual, thx to @typescript-bot for the heads up 👍 

I'm a bit busy with some unrelated topics. This pr got a bit messy. Going to close it, and reopen a new one soon.
